olp-cpp-sdk-core/tests/thread/unittest/SyncQueueTest.cpp

```diff
@@ -0,0 +1,207 @@
+/*
+ * Copyright (C) 2019 HERE Europe B.V.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ * License-Filename: LICENSE
+ */
+
+#include <gmock/gmock.h>
+#include <gtest/gtest.h>
+#include <atomic>
+#include <chrono>
+#include <thread>
+
+#include <olp/core/thread/SyncQueue.h>
+
+using SyncTaskType = std::function<void()>;
+using SyncQueueFifo = olp::thread::SyncQueueFifo<SyncTaskType>;
+
+using SyncIncType = std::function<void(size_t)>;
+using SyncQueueIncFifo = olp::thread::SyncQueueFifo<SyncIncType>;
+
+using SharedQueueType = std::shared_ptr<std::string>;
+using SyncQueueShared = olp::thread::SyncQueueFifo<SharedQueueType>;
+
+using namespace ::testing;
+using namespace std::chrono;
+
+TEST(SyncQueueTest, initialize) {
+  SCOPED_TRACE("Default initialized not closed");
+  SyncQueueFifo sync_queue;
+  bool check = false;
+  sync_queue.Push([&check]() { check = true; });
+
+  bool closed = false;
+  SyncTaskType element;
+  sync_queue.Pull(element, closed);
+  ASSERT_TRUE(element) << "Push() or Pull() failed.";
+  element();
+
+  EXPECT_TRUE(sync_queue.Empty());
+  EXPECT_FALSE(closed);
+  EXPECT_TRUE(check);
+}
+
+TEST(SyncQueueTest, push) {
```

Comment:
Divide into 2 tests - for rvalues and laves

Reply:
No need. With SCOPED_TRACE you can easily see what failed and you don't need to carry around so many test cases. This is very helpfull in case you have unit/integration tests with complex and cost-intensive setup/teardown.

Reply:
Disagree:

- this test has 2 cases, which do not depend on each other; So, it is logical to divide them into separate test cases;
- what if I want to run only one case (consider example when one case fails and other not) - in your approach, still 2 cases will run;
- currently in SynQueueTest there are no complex setup / teardown;
- we should strive to minimize assertion per test - ideally it should be 1 assertion per test.

Reply:
- This tests one method, as this is a method oriented test setup so it is more logical to merge them together as then you have an overview of what was tested and whatnot and you can easily merge test scenarios in one.
- What is the problem on running 2 cases in one? In most scenarios you can see based on previous scenarios, in the same test, if the test approach was wrong or you changed something you should have not.
- But consistency is key in shared projects. Imagine every test has a different paradigm based on developer, type of class, API, etc. You need a lot of time to adapt.
- Why? I didn't see any rule for this and this will be very hard to accomplish, resp we will have endless execution time on CI if we test each member of the class for each scenario in one test case. This will explode.

Reply:
- if this is method oriented test setup, then we should have 2 tests per methods, cause there are 2 push methods with different parameters;
- this more like structural split of test case; the problem is overhead when you want to run just one test case - 2 cases will be checked instead single one;
- we are relatively new team and should agree on how our tests should looks like; It is ok to discuss them now;
- Many TDD folks uses this approach. But it is not a mandatory rule, but nice to have; 
Detailed discussion with pros and cons on SE: https://softwareengineering.stackexchange.com/questions/7823/is-it-ok-to-have-multiple-asserts-in-a-single-unit-test

Reply:
There should be different tests against different scenarios. This allows to isolate a particular case for whichever reason. The simplest benefit is test name notification.
If a scenario requires a complex things to run in a queue/parallel this also should be a separate test case.
So, I agree with @borzun, this case should be split.